LIR definition and Cranelift backend

nbp opened the pull request:
This pull request is a work-in-progress to handle the LIR to Cranelift IR conversion. It adds a LIR representation #29 and adds a Cranelift backend #28 .

However it does not yet change the MIR to LIR mapping nor the serialization of the LIR. Thus, this pull request cannot yet be tested with a complete pipeline to handle the Brainfuck example. Tests for this pull requests are added as part of the test of the newly added `codegen` sub-project. These tests are creating LIR units which are then lowered into Cranelift IR, compiled and tested.

nbp committed:
Add a codegen sub-project to convert the LIR to machine code using Cranelift.

nbp committed:
Move NumberType and NumberValue to a dedicated module.

nbp committed:
Improve the LIR for the codegen test case.

nbp committed:
Extract Cranelift signature from the LIR.

nbp committed:
Add a Nix expression to remove extra nix-shell arguments.

nbp committed:
Replace the codegen add1 test case by one produced from the LIR.

nbp committed:
LIR: Add carry flag next to the overflow flag.

nbp committed:
codegen: Add a Convert context to hold the common state.

nbp committed:
codegen: Add a new test case to cover more LIR uses.

nbp committed:
Generate code for overflow and carry flags for additions.

nbp committed:
Add builders add_op_deps and corresponding end functions.

nbp committed:
Add builders the ability to know when we are done adding predecessors to a block.

nbp committed:
Fix end_ins instruction.

nbp committed:
Add SSA builder to automagically add Rehash & Phi instructions.

nbp committed:
Update Readme roadmap section with the latest milestone. (v01 and v0.2)

nbp committed:
Generate code for carry and overflow flag for substraction

nbp committed:
Move tests from condegen/src/lib.rs to codegen/tests directory.

nbp committed:
Fix Add/Sub overflow computation, and check results against rust library.

nbp committed:
Generate code for load and store instructions.

nbp committed:
Fix result type of Load and Store instructions.

nbp committed:
Records static references on the context and generate code for StaticAddress instructions.

nbp committed:
Generate code for calling function, with multiple arguments and return values.

nbp committed:
Add test cases to verify calls with multiple arguments and returned values.

nbp committed:
Generate code to handle conditions: Eq, Ne, Lt, Le, Gt, Ge and Ord.

nbp committed:
Generate code for Muliplication without overflow handling.

Hywan commented on codegen/src/lower.rs:
```diff
@@ -41,6 +41,70 @@ struct ConvertCtx<'a> {
+    use self::IFCond::*;
+    match op {
+        Ord(Ordered(t)) => {
+            assert_eq!(t.is_float(), true);
```
I think it is safe to write:

```rust
assert!(t.is_float());
```

instead of `assert_eq!`.

nbp replied:
Thanks! I will fix it next week-end.

If you have problem with some of the comments or the lack of comments, or think some descriptions are not clear, feel free to report them as well. I am trying to document as much as possible from the beginning and make these comments as readable/understandable as possible.

Hywan replied:
Everything is really great so far, you're doing an excellent job!

Hywan commented on lir/src/number.rs:
```diff
@@ -0,0 +1,156 @@
+/// can be convert to this one, including NumberValue, using the `into()`
+/// method.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum NumberType {
```
Is it interesting to support `i128` and `u128` here?

nbp replied:
The problem with having `NumberType` for `i128` and `u128` is that we need a corresponding `NumberValue` and at the moment `NumberValue` are part of the `Opcode::Const`, Thus increasing the size of all opcodes.

Support for 128 bits number is something which would have to be handled to support Rust fully, but this can be expressed as 64 bits operations.

Hywan replied:
Make sense! Thanks.

nbp committed:
Use assert! instead of assert_eq! when checking a precondition in opcode_to_cond

nbp committed:
Add StackAddress to reserve explicit stack slots in the generated code.

nbp committed:
Replace HolyJit lib own definition of the LIR and Compiler

nbp committed:
Change ComplexType to split the memory representation from the data stored in it.

nbp committed:
Add a readable Display implementation of the LIR format

nbp committed:
Add an unsafe interface to register statics.

The Library macro create an empty tuple pointer storage to reference the space
which contains the value of all the statics value sused during the compilation.

This implies that the library cannot use the untyped tuple type to derive the
size, using mem::size_of, to assert its size when registering the memory of
statics.

nbp committed:
Disable debug mode to reduce HolyJIT compilations.

The debug mode is enabling `-Z dump-mir=all` flag which attemps to spew the
content of the `&[u8]` array containing the serialized content. This arrau
serialization is time consuming and should be removed in the future by replacing
it by a `&str`.
